jobserver/views/users.py

```diff
@@ -1,82 +1,119 @@
-                # that they don't have a password to reset
+                # we don't want to expose users email address to a bad actor
+                # via the login page form so we're emailing them with a link
+                # to the login page to click the right button.
+                # We can't email them with a link to the social auth entrypoint
+                # because it only accepts POSTs.
                 send_github_login_email(user)
             else:
-                send_reset_password_email(user)
-
-        messages.success(
-            self.request, "Your password reset request was successfully sent."
-        )
-        return redirect("/")
-
-
-class SetPassword(PasswordResetConfirmView):
-    post_reset_login = True
-    post_reset_login_backend = "django.contrib.auth.backends.ModelBackend"
-    success_url = "/"
-    template_name = "set_password.html"
-
-    def get_success_url(self):
-        # the authorization framework doesn't support getting all roles for a
-        # user across all objects, so we hae to manually do it here.
-        roles = set(self.request.user.roles)
-        for membership in self.request.user.project_memberships.all():
-            roles |= set(membership.roles)
-
-        # treat users with only the InteractiveReporter role as interactive users.
-        # we'll have users with this role along with others, eg for staff, but
-        # they likely don't want to be redirected on login.
-        if roles != {InteractiveReporter}:
-            return "/"
-
-        # if the user is logging in via this view, so with an email address and
-        # password, _and_ has the InteractiveReporter role then they should
-        # definitely be an interactive user.
-
-        # TODO: add a page for users who have 2+ projects
-        project = self.request.user.projects.first()
-        return project.get_interactive_url()
+                self.request.session[INTERNAL_LOGIN_SESSION_TOKEN] = user.email
+                send_login_email(
+                    user, timeout_minutes=settings.LOGIN_URL_TIMEOUT_MINUTES
+                )
+
+        msg = "If you have a user account we'll send you an email with the login details shortly. If you don't receive an email please check your spam folder."
+        messages.success(self.request, msg)
+
+        context = self.get_context_data(next_url=self.next_url)
+        return self.render_to_response(context)
+
+    def get_context_data(self, **kwargs):
+        return super().get_context_data(**kwargs) | {
+            "next_url": self.next_url,
+        }
+
+    def get_next_url(self, request):
+        next_url = request.GET.get("next") or "/"
+        if not is_safe_path(next_url):
+            next_url = ""
+
+        return next_url
+
+
+class LoginWithURL(View):
+    def get(self, request, *args, **kwargs):
+        try:
+            email = request.session.pop(INTERNAL_LOGIN_SESSION_TOKEN)
```

Comment:
Similar to the above, we need to compare the unsigned url token to the stored session token.

If we don't do this, then an attacker could potentially use a signed token for any login request.

Also, we should probably only delete the session data on *successful* login, or else copy/pasting to the right browser won't work.

Altogether, I think this looks something like this:


```

try:
    email, token = request.session.get(INTERNAL_LOGIN_SESSION_TOKEN)
    url_token = TimestampSigner(salt="login").unsign(
        self.kwargs["token"],
        max_age=timedelta(minutes=settings.LOGIN_URL_TIMEOUT_MINUTES),
    )
    if url_token != token:
        raise BadToken()
    user = User.objects.get(email=email)
    request.session.pop(INTERNAL_LOGIN_SESSION_TOKEN)
except (BadSignature, BadToken, KeyError, SignatureExpired, User.DoesNotExist) as e:
    ...
    
```